Local WebKit builds don't use locally-built WebKit process extensions

pvollan opened the pull request:
#### 314481e35106de8ac716b2da7838e3760b54eadc
<pre>
Local WebKit builds don&apos;t use locally-built WebKit process extensions
<a href="https://bugs.webkit.org/show_bug.cgi?id=273550">https://bugs.webkit.org/show_bug.cgi?id=273550</a>
<a href="https://rdar.apple.com/125726458">rdar://125726458</a>

Reviewed by Elliott Williams.

Add Xcode script build phase to copy extensions to app bundle.

* Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm:
(WebKit::isInWebKitChildProcess):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::hasExtensionsInAppBundle):
(WebKit::launchWithExtensionKit):
* Tools/WebKitTestRunner/Configurations/WebKitTestRunnerApp.xcconfig:
* Tools/WebKitTestRunner/Configurations/WebKitTestRunnerApp-iOS.entitlements:
* Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/279665@main">https://commits.webkit.org/279665@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5cfdd630dcda7b8a28f71f5dd247ff7c03d2d4c9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33377 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6535 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57280 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4729 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56306 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40890 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4620 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43724 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3126 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56101 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31596 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46733 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24865 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28423 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4052 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2878 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/50088 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4253 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58874 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29186 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4340 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51141 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30370 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46860 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50484 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31325 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8018 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30151 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple reviewed (commented):
As-is, the script phase needs input and output paths defined. Otherwise this script will run on every rebuild, add log clutter, and slow down incremental builds.

But I don't like doing copying in a script phase at all, it's too easy to make a mistake and break incremental builds. If possible, you should use a Copy Files phase for this, and move the entitlement and Info.plist logic to the appex targets themselves.

emw-apple commented on Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/project.pbxproj:
```diff
@@ -174,6 +177,22 @@
+			outputFiles = (
+				"$(BUILT_PRODUCTS_DIR)/WebKitTestRunnerApp.app/Extensions/$(INPUT_FILE_NAME)",
+			);
+			script = "if [[ ${WK_PLATFORM_NAME} == \"macosx\" ]]; then\n    exit\nfi\n\nDESTINATION_PATH=${BUILT_PRODUCTS_DIR}/WebKitTestRunnerApp.app/Extensions/${INPUT_FILE_NAME}\nditto ${INPUT_FILE_PATH} ${DESTINATION_PATH}\n\n/usr/libexec/PlistBuddy -c \"Set :CFBundleIdentifier ${PRODUCT_BUNDLE_IDENTIFIER}.${INPUT_FILE_BASE}\" \"${DESTINATION_PATH}/Info.plist\"\n/usr/libexec/PlistBuddy -c \"Delete :EX ...
```
Two small suggestions:

- Add `set -e` to the very beginning, so that failures from `ditto` or PlistBuddy stop the build.
- Remove the `if [[ ${WK_PLATFORM_NAME} == "macosx" ]]; then exit` clause. Exiting without creating the output means we're lying to the build system, because it expects the rule to produce this output unconditionally. Instead, you can add `*.appex` to EXCLUDED_SOURCE_FILE_NAMES here https://github.com/WebKit/WebKit/blob/4d53fb735617548d3f4402459fb426c20d232c1d/Tools/WebKitTestRunner/Configurations/WebKitTestRunnerApp.xcconfig#L40, and that will prevent it from scheduling the rule at all.

pvollan replied:
Ah, good points! I will address these issues in the upcoming patch.

Thanks for reviewing!

emw-apple commented on Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/project.pbxproj:
```diff
@@ -174,6 +177,22 @@
+			);
+			isEditable = 1;
+			outputFiles = (
+				"$(BUILT_PRODUCTS_DIR)/WebKitTestRunnerApp.app/Extensions/$(INPUT_FILE_NAME)",
```
```suggestion
				"$(BUILT_PRODUCTS_DIR)/$(EXTENSIONS_FOLDER_PATH)/$(INPUT_FILE_NAME)",
```

emw-apple commented on Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/project.pbxproj:
```diff
@@ -231,6 +250,18 @@
 			name = "Copy Plug-Ins";
 			runOnlyForDeploymentPostprocessing = 0;
 		};
+		E372BC342C08E22D006DFE67 /* CopyFiles */ = {
```
Could you give this a name, maybe "Embed extensions"? (In Xcode, click on the "Copy Files" label and it will magically become a text field!)

pvollan replied:
Yes, good point, will do!

emw-apple reviewed (commented):
Looks good! Couple of minor suggestions, and then I'm ready to r+ this.

pvollan committed:
Local WebKit builds don't use locally-built WebKit process extensions
https://bugs.webkit.org/show_bug.cgi?id=273550
rdar://125726458

Reviewed by Elliott Williams.

Add Xcode script build phase to copy extensions to app bundle.

* Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm:
(WebKit::isInWebKitChildProcess):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::hasExtensionsInAppBundle):
(WebKit::launchWithExtensionKit):
* Tools/WebKitTestRunner/Configurations/WebKitTestRunnerApp.xcconfig:
* Tools/WebKitTestRunner/Configurations/WebKitTestRunnerApp-iOS.entitlements:
* Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/279665@main
